### ISSUE TITLE
Emptying siblings is possible when creating new atom types; use hierarchy to avoid this

### ISSUE THREAD

**Issue by davidlmobley** [open-forcefield-group/smarty#77]
Currently, creation of a new child atom type is not allowed to empty parent atom types. However, there are no checks in place to ensure creation of a child atom type does not empty SIBLING atom types. So currently in AlkEthOH at T=0, it is possible to end up with this tree:

```
Atom type hierarchy:
    [#1]
        [#1$(*~[#6]~[#1])]
            [#1$(*~[#6](-[#8])~[#1])]
            [#1$(*~[#6](~[#8])~[#1])]
        [#1$(*-[#8])]
    [#8]
        [#8$(*~[#1])]
    [#6]
```

because creation of the second `H1`-equivalent SMARTS pattern empties its sibling without emtpying its parents. Specifically: 

```
INDEX        ATOMS  MOLECULES                                                          TYPE NAME                           SMARTS REF TYPE        FRACTION OF REF TYPED MOLECULES MATCHED
    1 :         68         42 |                                                       c_hydrogen                             [#1]       H2               27 /               33 ( 81.818%)
    2 :        232         42 |                                                         c_carbon                             [#6]       CT              232 /              232 (100.000%)
    3 :         39         30 |                                                         c_oxygen                             [#8]       OS               39 /               39 (100.000%)
    4 :         68         42 |                                       c_oxygen hydrogen-adjacent                    [#8$(*~[#1])]       OH               68 /               68 (100.000%)
    5 :        244         38 |                           c_hydrogen any c_carbon any c_hydrogen               [#1$(*~[#6]~[#1])]       HC              244 /              244 (100.000%)
    6 :         68         42 |                                      c_hydrogen simply c_oxygen                     [#1$(*-[#8])]       HO               68 /               68 (100.000%)
    7 :          0          0 |         c_hydrogen any c_carbon any c_hydrogen (simply c_oxygen)        [#1$(*~[#6](-[#8])~[#1])]
    8 :         84         25 |            c_hydrogen any c_carbon any c_hydrogen (any c_oxygen)        [#1$(*~[#6](~[#8])~[#1])]       H1               78 /              116 ( 67.241%)
TOTAL :        803         42 |                                                                                                         756 /      803 match (94.147 %)
```

I don't think this is a particularly high priority to fix because we have larger issues (needing to go to finite temperature to be able to backtrack from dead ends, in particular) but I'm noting it for the record. We should fix at some point by checking to ensure creation of new children doesn't empty their siblings.

**Comment by jchodera** [comment]
We should probably just be checking to make sure no atom type has zero matches after a proposal. I _thought_ we did this before, but I guess I must be wrong!

**Comment by davidlmobley** [comment]
> We should probably just be checking to make sure no atom type has zero matches after a proposal. I thought we did this before, but I guess I must be wrong!

It's probably unnecessary to check that NO atom type has zero matches (though that's indeed the simplest thing to check). And no, we didn't do this before that I recall - we just checked that parents didn't end up with no matches (which is indeed the thing which happens most often).

**Comment by bannanc** [comment]
@davidlmobley or @camizanette Did you change this in smarty? I'm definitely running into this issue with smirky now, I made a separate issue before remembering David identified it before.

**Comment by davidlmobley** [comment]
Not fixed, @bannanc ; if you want to put in a fix that'd be great or you can ask Camila to look at it.

**Comment by bannanc** [comment]
@camizanette (and @davidlmobley ) given our current divide of labor I'm going to try addressing the issue in smirky first then if I get it to work I'll see if the same fix works on smarty.

**Comment by davidlmobley** [comment]
Sounds perfect, @bannanc .

**Comment by bannanc** [comment]
@davidlmobley - I found the issue, I fixed it on smirky and then didn't come back to fixing smarty.

**Comment by davidlmobley** [comment]
Ah, right, thanks, @bannanc .

**Comment by bannanc** [comment]
fixed in pull request #198
